// @node-projects/web-component-designer-zpl/dist/widgets/views/zpl-demo-view.js
import { BaseCustomWebComponentConstructorAppend, css, html } from '@node-projects/base-custom-webcomponent';
export class ZplDemoView extends BaseCustomWebComponentConstructorAppend {
    static template = html `<h2>Label generated via http://api.labelary.com/</h2><br><img id="image">`;
    static style = css `
        :host {
            display: block;
            overflow: hidden;
            background: white;
            height: 100%;
            width: 100%;
            position: relative;
        }
        #image {
            border: solid 1px black;
        }`;
    constructor() {
        super();
    }
    executeCommand;
    canExecuteCommand;
    dispose() { }
    async display(serviceContainer, instanceServiceContainer, code, style) {
        const width = 4;
        const height = 6;
        const dpmm = '24dpmm';
        const response = await fetch(`https://api.labelary.com/v1/printers/${dpmm}/labels/${width}x${height}/${0}`, {
            method: "POST",
            body: code,
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded',
                'Accept': 'image/png'
            },
        });
        this._getDomElement('image').src = URL.createObjectURL(await response.blob());
        ;
    }
}
customElements.define('node-projects-zpl-demo-view', ZplDemoView);
